contrib/base64-cmake/CMakeLists.txt

```diff
@@ -11,50 +11,74 @@ endif()
-add_library(_base64_ssse3 OBJECT "${LIBRARY_DIR}/turbob64sse.c") # This file also contains code for ARM NEON
-
-if (ARCH_AMD64)
-    add_library(_base64_avx OBJECT "${LIBRARY_DIR}/turbob64sse.c") # This is not a mistake. One file is compiled twice.
-    add_library(_base64_avx2 OBJECT "${LIBRARY_DIR}/turbob64avx2.c")
+set(HAVE_NEON32 0)
+set(HAVE_NEON64 0)
+if (ARCH_AARCH64)
+    set(HAVE_NEON64 1)
 endif ()
 
-target_compile_options(_base64_scalar PRIVATE -falign-loops)
-
-if (ARCH_AMD64)
-    target_compile_options(_base64_ssse3 PRIVATE -mno-avx -mno-avx2 -mssse3 -falign-loops)
-    target_compile_options(_base64_avx PRIVATE -falign-loops -mavx)
-    target_compile_options(_base64_avx2 PRIVATE -falign-loops -mavx2)
-else ()
-    if (ARCH_PPC64LE)
-        target_compile_options(_base64_ssse3 PRIVATE -D__SSSE3__ -falign-loops)
+macro(cast_to_bool var instruction)
+    if (HAVE_${var})
+        set(base64_${var} 1)
+        set(base64_${var}_opt ${instruction})
     else()
-        target_compile_options(_base64_ssse3 PRIVATE -falign-loops)
+        set(base64_${var} 0)
     endif()
-endif ()
+endmacro()
+
+cast_to_bool(SSSE3 "-mssse3")
+cast_to_bool(SSE41 "-msse4.1")
+cast_to_bool(SSE42 "-msse4.2")
+cast_to_bool(AVX   "-mavx")
+cast_to_bool(AVX2  "-mavx2")
 
-if (ARCH_AMD64)
-    add_library(_base64
-        $<TARGET_OBJECTS:_base64_scalar>
-        $<TARGET_OBJECTS:_base64_ssse3>
-        $<TARGET_OBJECTS:_base64_avx>
-        $<TARGET_OBJECTS:_base64_avx2>)
-else ()
-    add_library(_base64
-        $<TARGET_OBJECTS:_base64_scalar>
-        $<TARGET_OBJECTS:_base64_ssse3>)
+set(HAVE_FAST_UNALIGNED_ACCESS 0)
+if(HAVE_SSSE3 OR HAVE_SSE41 OR HAVE_SSE42 OR HAVE_AVX OR HAVE_AVX2 OR ARCH_AARCH64)
+    set(HAVE_FAST_UNALIGNED_ACCESS 1)
 endif ()
 
-target_include_directories(_base64 SYSTEM PUBLIC ${LIBRARY_DIR})
+configure_file(config.h.in ${CMAKE_CURRENT_BINARY_DIR}/config.h)
 
-if (XCODE OR XCODE_VERSION)
-    # https://gitlab.kitware.com/cmake/cmake/issues/17457
-    #     Some native build systems may not like targets that have only object files, so consider adding at least one real source file
-    # This applies to Xcode.
-    if (NOT EXISTS "${CMAKE_CURRENT_BINARY_DIR}/dummy.c")
-        file(WRITE "${CMAKE_CURRENT_BINARY_DIR}/dummy.c" "")
-    endif ()
-    target_sources(_base64 PRIVATE "${CMAKE_CURRENT_BINARY_DIR}/dummy.c")
-endif ()
+add_library(_base64
+        # Library files
+        "${LIBRARY_DIR}/lib/lib.c"
+        "${LIBRARY_DIR}/lib/codec_choose.c"
+
+        "${LIBRARY_DIR}/lib/tables/tables.c"
+
+        # Codec implementations
+        "${LIBRARY_DIR}/lib/arch/generic/codec.c"
+
+        "${LIBRARY_DIR}/lib/arch/ssse3/codec.c"
+        "${LIBRARY_DIR}/lib/arch/sse41/codec.c"
+        "${LIBRARY_DIR}/lib/arch/sse42/codec.c"
+        "${LIBRARY_DIR}/lib/arch/avx/codec.c"
+        "${LIBRARY_DIR}/lib/arch/avx2/codec.c"
+        "${LIBRARY_DIR}/lib/arch/avx512/codec.c"
```

Comment:
Building avx512 is pointless if we don't define `HAVE_AVX512` (l. 34).

Reply:
We can't simply omit this file unfortunately, because there's a possible run-time codec choose, which requires the function to be present on link time: https://github.com/aklomp/base64/blob/master/lib/codec_choose.c#L146-L150

If we want to drop this, we'll have to patch it out.

Reply:
Then keep it in, but kindly add a comment (basically what you wrote before) so that readers are not confused, thanks!